### PR TITLE
Solid fuel (coal, oats, wood) support for vehicle engines

### PR DESCRIPTION
**Pull request by mlangsdorf**
#### Summary
```SUMMARY: Features "Add support for solid fuels like coal for vehicle engines"```

#### Purpose of change
Allow people to use vehicle engines that consume solid matter, such as coal burning steam engines.
Mod writers will be able to use this to make engines powered by just about anything.

#### Describe the solution
There were a lot of assumptions in the vehicle code that fuel was either electricity coming from a battery
or liquids coming from a tank. I removed a lot of those assumptions, and now any vehicle part with a
MAGAZINE as its base part can be used to store fuel that will be consumed by the engine.

I've also added a sample coal burning, closed cycle, steam engine.

#### Describe alternatives you've considered
This was pretty straightforward.

#### Additional content
I added 125 HP and 200 HP steam engines to the normal spawn lists. Larger steam engines and more
compact steam turbines were added to blazemod but don't currently spawn anywhere.

### CODE REVIEW

File: data/json/items/vehicle/engine.json [hunk 1]
```diff
@@ -39,6 +39,21 @@
...
+    "abstract": "engine_steam",
+    "copy-from": "engine_combustion",
+    "type": "ENGINE",
+    "name": "base steam engine",
+    "looks_like": "v12_diesel",
+    "faults": [
+      "fault_engine_belt_drive",
+      "fault_engine_filter_air",
+      "fault_engine_filter_fuel",
+      "fault_engine_immobiliser",
```
Comment (Night-Pryanik) [inline data/json/items/vehicle/engine.json]: Immobilizer on a steam engine?

File: data/json/items/vehicle/engine.json [hunk 2]
```diff
@@ -39,6 +39,21 @@
...
+  {
+    "abstract": "engine_steam",
+    "copy-from": "engine_combustion",
+    "type": "ENGINE",
+    "name": "base steam engine",
+    "looks_like": "v12_diesel",
+    "faults": [
+      "fault_engine_belt_drive",
+      "fault_engine_filter_air",
+      "fault_engine_filter_fuel",
```
Comment (Night-Pryanik) [inline data/json/items/vehicle/engine.json]: I don't know much about steam engines, but is there actually a fuel filter for solid fuel like coal?

Comment (mlangsdorf) [reply data/json/items/vehicle/engine.json]: Probably not!

File: data/json/items/vehicle/engine.json (hunk 1 above)
Comment (mlangsdorf) [reply data/json/items/vehicle/engine.json]: Will fix.

Comment (mlangsdorf) [reply data/json/items/vehicle/engine.json]: Nonsense faults removed.

File: data/json/items/vehicle/engine.json (hunk 2 above)
Comment (mlangsdorf) [reply data/json/items/vehicle/engine.json]: Nonsense faults removed.

**Comment by kwl01skz** [comment]
Cool.
This is a very reasonable behavior when resources are scarce.

File: data/json/recipes/recipe_vehicle.json [hunk 3]
```diff
@@ -1,4 +1,32 @@
 [
+  {
+    "result": "rebar_plate",
```
Comment (Reclusive-reptile) [inline data/json/recipes/recipe_vehicle.json]: Is this right?

Comment (mlangsdorf) [reply data/json/recipes/recipe_vehicle.json]: Nope, it should be the makeshift steam engine. Thanks for the review.

Comment (mlangsdorf) [reply data/json/recipes/recipe_vehicle.json]: Fixed.

**Comment by DegenerateGroundState** [comment]
Reminds me of wood gasifiers, often used in reality (in internal combustion engines) when petroleum products are unavailable
https://en.wikipedia.org/wiki/Wood_gas

**Comment by mlangsdorf** [comment]
I want to support gassifiers, because they are a real thing, but I don't want to write the necessary support for CDDA to handle gasses correctly.

Wood burning steam engines are going to be part of the flexfuel update Real Soon Now.

**Comment by ZhilkinSerg** [comment]
You can refill fuel bunker with anything:

image

Tried plastic chunks:

image

**Comment by mlangsdorf** [comment]
Okay, I'll go fix that.

**Comment by ZhilkinSerg** [comment]
1. Fuel unload is instant action which takes no time.
2. Steam engines seems to be a little overpowered (i.e. 350 kph speed for primitive tractor with medium steam engine off-road), but it is not something introduced in this PR solely and affects other engines too.

Otherwise looks good in game.

**Comment by mlangsdorf** [comment]
Engine | HP | Mass (kg) | HP/kg | Fuel GJ/vehicle tile | Fuel Mass/Tile
------| -- | --------- | ----- | ------- | ----------
makeshift steam | 75 | 225 | .33 | 150 | 88
gas i4 | 125 | 130 | .95 | 205 | 60
small steam | 125 | 180 | .64 | 150 | 88
diesel v6 | 200 | 180 | 1.1 | 215 | 60
gas v6 | 200 |  190 | 1.05 | 205 | 60
medium steam | 200 | 280 | .72 | 150 | 88

Steam engines are louder and heavier per HP than gas or diesel and burn fuel that gets less mileage per volume while weighing more. C:DDA vehicles go too fast for their nominal HP, but this PR doesn't introduce that.

I'll look at the instant unload issue. I hate writing player activities.

**Comment by ZhilkinSerg** [comment]
> C:DDA vehicles go too fast for their nominal HP, but this PR doesn't introduce that.

I agree - engines were overpowered long before this PR and are subject to change sometime later.

> I'll look at the instant unload issue. I hate writing player activities.

This one could be addressed in separate PR later as it is not a big deal right now - you instantly receive huge pile of coal which then should be handled using time.

File: src/vehicle.h [hunk 4]
```diff
@@ -222,6 +222,11 @@ struct vehicle_part {
         /** Is this any type of vehicle light? */
         bool is_light() const;
 
+        /** Can this part store fuel of any type
+         * @no_broke exclude broken parts
+         */
+        bool is_fuel_store( const bool no_broke = true ) const;
```
[the remark this answers is not in the document]
Comment (mlangsdorf) [reply src/vehicle.h]: The C++ standard doesn't require they be the same but I will fix in the morning when I fix the merge issue.

[the remark this answers is not in the document]
Comment (mlangsdorf) [reply src/vehicle.h]: Fixed.

**Pull request merged**